src/SimplifyCorrelatedDifferences.cpp

```diff
@@ -125,16 +199,28 @@ class SimplifyCorrelatedDifferences : public IRMutator {
             (ma == Monotonic::Decreasing && mb == Monotonic::Increasing && std::is_same<T, Add>::value)) {
 
             for (auto it = lets.rbegin(); it != lets.rend(); it++) {
-                e = Let::make(it->first, it->second, e);
+                if (expr_uses_var(e, it->name)) {
+                    if (!it->may_substitute) {
+                        // We have to stop here. Can't continue
+                        // because there might be an outer let with
+                        // the same name that we *can* substitute in,
+                        // and then inner uses will get the wrong
+                        // value.
+                        break;
+                    }
+                }
+                e = Let::make(it->name, it->value, e);
             }
             e = common_subexpression_elimination(e);
             e = solve_expression(e, loop_var).result;
+            e = PartiallyCancelDifferences().mutate(e);
             e = simplify(e);
 
             if ((debug::debug_level() > 0) &&
-                is_monotonic(e, loop_var, monotonic) == Monotonic::Unknown) {
+                is_monotonic(e, loop_var) == Monotonic::Unknown) {
                 // Might be a missed simplification opportunity. Log to help improve the simplifier.
```

Comment:
Suggestion: add an Issue for "things we could log in a structured way during batch compiles to gather useful data", and add a TODO for it here